sweep/sweeper.go

```diff
@@ -1521,12 +1393,76 @@ func (s *UtxoSweeper) handleNewInput(input *sweepInputMessage,
 	)
 	if err != nil {
 		err := fmt.Errorf("wait for spend: %w", err)
-		s.signalAndRemove(&outpoint, Result{Err: err})
+		s.markInputFailed(pi, err)
 
 		return
 	}
 
-	pendInput.ntfnRegCancel = cancel
+	pi.ntfnRegCancel = cancel
+}
+
+// decideStateAndRBFInfo queries the mempool to see whether the given input has
+// already been spent. If so, the state Published will be returned, otherwise
+// state Init. When spent, it will query the sweeper store to fetch the fee
+// info of the spending transction, and construct an RBFInfo based on it.
+// Suppose an error occurs, fn.None is returned.
+func (s *UtxoSweeper) decideStateAndRBFInfo(op wire.OutPoint) (
+	SweepState, fn.Option[RBFInfo]) {
+
+	// Check if we can find the spending tx of this input in mempool.
+	txOption := s.mempoolLookup(op)
+
+	// Extract the spending tx from the option.
+	var tx *wire.MsgTx
+	txOption.WhenSome(func(t wire.MsgTx) {
+		tx = &t
+	})
+
+	// Exit early if it's not found.
+	//
+	// NOTE: this is not accurate for backends don't support mempool
+	// lookup:
+	// - for neutrino we don't have a mempool.
+	// - for bitcoind below v24.0.0 we don't have `gettxspendingprevout`.
+	// - for btcd below v0.24.1 we don't have `gettxspendingprevout`.
+	if tx == nil {
+		return StateInit, fn.None[RBFInfo]()
+	}
+
+	// Otherwise the input is already spent in the mempool, so eventually
+	// we will return StatePublished.
+	//
+	// We also need to update the RBF info for this input. If the sweeping
+	// transaction is broadcast by us, we can find the fee info in the
+	// sweeper store.
+	txid := tx.TxHash()
+	tr, err := s.cfg.Store.GetTx(txid)
```

Comment:
hmmm not sure if I understand it correctly, are you saying the stored tx can be different from the tx fetched from the mempool?

Reply:
> Perhaps we should determine the RBF info directly from the tx instead.

Deriving the fee information requires access to all the previous inputs to have access to the total input sum. So we'd need to iteratively fetch the transactions of all the referenced inputs, but some of those might already be confirmed, so we can't get it solely from the mempool.

Reply:
Exactly imagine the case where and HTLC is swept by the preimage after we registered it with the sweeper. But this preimage monitoring is already handled in the resolver. Not sure if we wanna open this discussion here because we immediately fails the incoming back as soon as we see the preimage in mempool